src/ml/clustering/kmeans.nim

```diff
@@ -0,0 +1,268 @@
+# Copyright (c) 2018 Mamy André-Ratsimbazafy and the Arraymancer contributors
+# Distributed under the Apache v2 License (license terms are at http://www.apache.org/licenses/LICENSE-2.0).
+# This file may not be copied, modified, or distributed except according to those terms.
+import math, random, tables
+
+import
+  ../../tensor/tensor, ../../linear_algebra/linear_algebra
+
+proc euclidean_distance [T: SomeReal](u: Tensor[T], v: Tensor[T], squared: bool = false): T =
+  ## Calculates the euclidean distance
+  ## Inputs:
+  ##  - u: a tensor
+  ##  - v: a tensor
+  ##  - squared: whether or not to square the result
+  ## sqrt( (u - v) dot (u - v) )
+  ## which can be reformulated
+  ## sqrt((u dot u) - 2 * (u dot v) + (v dot v))
+  var u_v = (u .- v).reshape(u.shape[1])
+  result = dot(u_v, u_v)
+  if not squared:
+    result = sqrt(result) 
+
+proc cumsum [T: SomeReal](p: Tensor[T]): Tensor[T] =
+  ## Calculates the cumulative sum of a vector.
+  ## Inputs:
+  ##  - p: the tensor to cumulatively sum
+  assert p.rank == 1
+  assert p.shape[1] == 0
+  var n_rows = p.shape[0]
+  result = p
+  for i in 1..<n_rows:
+    result[i] += result[i-1]
+
+proc init_random [T: SomeReal](x: Tensor[T], n_clusters: int): Tensor[T] =
+  ## Helper method to randomly assign the initial centroids
+  result = newTensor[T](n_clusters, x.shape[1])
+  for i in 0..<n_clusters:
+    var random_point = rand(x.shape[0])
+    result[i, _] = x[random_point, _]
+
+proc get_closest_centroid [T: SomeReal](x: Tensor[T], centroids: Tensor[T], cid: int): int =
+  ## Helper method to get the closest centroid
+  var
+    closest_dist = Inf
+    closest_centroid_id: int
+  for closest in 0..<cid:
+    if euclidean_distance(x[cid, _], centroids[closest, _], squared = true) < closest_dist:
+      closest_centroid_id = closest
+  return closest_centroid_id
+
+proc get_candidates [T: SomeReal](n: int, distances: Tensor[T]): Tensor[int] =
+  ## Sample candidates with probability weighted by the distances
+  var
+    probs = cumsum(distances ./ @[distances.sum].toTensor)
```

Comment:
I tired removing the additional tensor related code for the denominator, but it fails because of a type mismatch:

e.g. 

```
probs = cumsum(distances ./ distances.sum)

proc `./`[T: SomeInteger](a, b: Tensor[T]): Tensor[T]
  first type mismatch at position: 1
  required type: Tensor[./.T]
  but expression 'distances' is of type: Tensor[system.float64]
proc `./`[T: SomeFloat](val: T; t: Tensor[T]): Tensor[T]
  first type mismatch at position: 1
  required type: T: SomeFloat
  but expression 'distances' is of type: Tensor[system.float64]
proc `./`[T: SomeInteger](val: T; t: Tensor[T]): Tensor[T]
  first type mismatch at position: 1
  required type: T: SomeInteger
  but expression 'distances' is of type: Tensor[system.float64]
proc `./`[T: SomeFloat](a, b: Tensor[T]): Tensor[T]
  first type mismatch at position: 2
  required type: Tensor[./.T]
  but expression 'sum(distances)' is of type: float64

expression: distances ./ sum(distances)
```

I think `distances ./ distances` but that would just reduce each value to 1. I switched the var to let.